### PR TITLE
8303509: Socket setTrafficClass does not work for IPv4 connections when IPv6 enabled

### PR DESCRIPTION
**Pull request by AlanBateman**
This is a small update to the underlying implementation of Socket.setTrafficClass so that it attempts to set the IP_TOS socket option when the socket is IPv6 and the operating system allows this option to be set on IPv6 sockets. The change is deliberately limited to the SocketImpl and SocketChannel implementation to avoid impacting DatagramChannel.

Socket.setTrafficClass is specified as a hint. For TCP connections it is possible on some platforms to set IP_TOS on an unbound socket to the ToS value to use in the SYN packet when establishing a connection. Setting it after a connection is established is possible on some platforms too. The JDK uses IPv6 sockets since JDK 1.4 so it gets a bit more complicated due to patchy support for IP_TOS.

- Linux allows IP_TOS to be set on an IPv6 socket at any time. If set on an unbound socket then the ToS will have the expected value when establishing a connection and it can be changed any time after connecting.
- macOS doesn't allow IP_TOS to be set on an IPv6 socket and it doesn't use the IPV6_TCLASS value as the IPv4 ToS. So on macOS, you have to use an IPv4 socket and set IP_TOS before connecting the socket.
- Windows hasn't allowed applications change the ToS for a long time. It allows the IP_TOS to be set on an unbound IPv6 socket but the value is not used.

The bug report report is Socket.setTrafficClass hasn't worked on Linux since JDK 13 when the SocketImpl was replaced. The workaround is to run with-Djava.net.preferIPv4=true. 

The proposed change is to just attempt to set both IPV6_TCLASS and IP_TOS when the socket is IPv6 and the operating system allows IPPROTO level sockets to be setup on IPv6 sockets. At some point I'd like to replace the underlying socket options implementation as it no longer needs to be in libnet, and that would be the limit to re-visit the equivalent in the channel implementations.

I had to use tcpdump and strace to test the changes, I don't think it's possible to create an automated test for this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303509](https://bugs.openjdk.org/browse/JDK-8303509): Socket setTrafficClass does not work for IPv4 connections when IPv6 enabled

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - Committer) ⚠️ Review applies to [c6766a4d](https://git.openjdk.org/jdk/pull/12872/files/c6766a4d971dc056cabbf270b1eda8162c62f13e)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12872/head:pull/12872` \
`$ git checkout pull/12872`

Update a local copy of the PR: \
`$ git checkout pull/12872` \
`$ git pull https://git.openjdk.org/jdk pull/12872/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12872`

View PR using the GUI difftool: \
`$ git pr show -t 12872`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12872.diff">https://git.openjdk.org/jdk/pull/12872.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
> Have you considered merging the new `setIpSocketOption` with the existing `DatagramChannelImpl.setOption`?

Yes, but not for this PR because there is a larger piece of work required here to get rid of a lot of cruft related to socket options.

**Review by Michael-Mc-Mahon — APPROVED** [review]
The change looks fine (with or without comment suggestion)

File: src/java.base/share/classes/sun/nio/ch/Net.java [hunk 1]
```diff
@@ -100,8 +99,8 @@ static boolean useExclusiveBind() {
     /**
      * Tells whether both IPV6_XXX and IP_XXX socket options should be set on
      * IPv6 sockets. On some kernels, both IPV6_XXX and IP_XXX socket options
-     * need to be set so that the settings are effective for IPv4 multicast
-     * datagrams sent using the socket.
+     * need to be set so that the settings are effective for IPv4 connections
+     * and datagrams.
      */
```
[the remark this answers is not in the document]
Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/Net.java:104]: Good point, this comments in the native implementations should be aligned.

**Review by Michael-Mc-Mahon — APPROVED** [review]
Comment change looks good.

**Pull request closed without merging**